transport/src/main/java/io/netty/channel/CoalescingBufferQueue.java

```diff
@@ -0,0 +1,225 @@
+     */
+    public void add(ByteBuf buf) {
+        ObjectUtil.checkNotNull(buf, "buf");
+        if (readableBytes > Integer.MAX_VALUE - buf.readableBytes()) {
+            throw new IllegalStateException("buffer queue length overflow: " + readableBytes
+                    + " + " + buf.readableBytes());
+        }
+        bufAndListenerPairs.add(buf);
+        readableBytes += buf.readableBytes();
+    }
+
+    /**
+     * Add a buffer to the end of the queue and associate a promise with it that should be completed when
+     * all the buffers bytes have been consumed from the queue and written.
+     * @param buf to add to the tail of the queue
+     * @param promise to complete when all the bytes have been consumed and written, can be void.
+     */
+    public void add(ByteBuf buf, ChannelPromise promise) {
+        // buffers are added before promises so that we naturally 'consume' the entire buffer during removal
+        // before we complete it's promise.
+        ObjectUtil.checkNotNull(promise, "promise");
+        if (promise.isVoid()) {
+            add(buf);
+        } else {
+            add(buf, new ChannelPromiseNotifier(promise));
+        }
+    }
+
+    /**
+     * Add a buffer to the end of the queue and associate a listener with it that should be completed when
+     * all the buffers  bytes have been consumed from the queue and written.
+     * @param buf to add to the tail of the queue
+     * @param listener to notify when all the bytes have been consumed and written.
+     */
+    public void add(ByteBuf buf, ChannelFutureListener listener) {
+        // buffers are added before promises so that we naturally 'consume' the entire buffer during removal
+        // before we complete it's promise.
+        ObjectUtil.checkNotNull(buf, "buf");
+        ObjectUtil.checkNotNull(listener, "listener");
+        if (readableBytes > Integer.MAX_VALUE - buf.readableBytes()) {
+            throw new IllegalStateException("buffer queue length overflow: " + readableBytes
+                    + " + " + buf.readableBytes());
+        }
+        bufAndListenerPairs.add(buf);
+        bufAndListenerPairs.add(listener);
+        readableBytes += buf.readableBytes();
+    }
+
+
+    /**
+     * Remove a {@link ByteBuf} from the queue with the specified number of bytes. Any added buffer who's bytes are
+     * fully consumed during removal will have it's promise completed when the passed aggregate {@link ChannelPromise}
+     * completes.
+     *
+     * @param bytes the maximum number of readable bytes in the returned {@link ByteBuf}, if {@code bytes} is greater
+     *              than {@link #readableBytes} then a buffer of length {@link #readableBytes} is returned.
+     * @param aggregatePromise used to aggregate the promises and listeners for the constituent buffers.
+     * @return a {@link ByteBuf} composed of the enqueued buffers.
+     */
+    public ByteBuf remove(int bytes, ChannelPromise aggregatePromise) {
```

Comment:
Is this more valuable than just supporting ChannelFutureListener? A quick scan of the Netty codebase shows that most calls to promise.isDone/isSuccess/... come from inside ChannelFutureListener implementations. Or to put it another way, people rarely poll promises, they just listen to them.

If we just accept ChannelFutureListeners here then we are supporting the majority use case cleanly. If someone wants to poll a promise then they can wrap it in a notifier themselves before adding it. Immediate completion is not an issue here